orc8r/cloud/go/swagger/combine.go

```diff
@@ -0,0 +1,197 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package swagger
+
+import (
+	"regexp"
+	"sort"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+// Combine multiple Swagger specs, giving precedence to the "common" spec.
+// First error contains merge warnings for overwritten fields and
+// incompatible Swagger specs.
+// This custom-built functionality mirrors the "official" implementation.
+// See: https://github.com/go-openapi/analysis/blob/master/mixin.go
+func Combine(yamlCommon string, yamlSpecs []string) (string, error, error) {
+	warnings := &multierror.Error{}
+
+	common, specs, errs := unmarshalToSwagger(yamlCommon, yamlSpecs)
+	if errs != nil {
+		multierror.Append(warnings, errs)
+	}
+
+	combined, errs := combine(common, specs)
+	if errs != nil {
+		multierror.Append(warnings, errs)
+	}
+
+	out, err := marshalToYAML(combined)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return out, warnings, nil
+}
+
+func combine(common Spec, specs []Spec) (Spec, error) {
+	var tags [][]TagDefinition
+	var paths []map[string]interface{}
+	var responses []map[string]interface{}
+	var parameters []map[string]interface{}
+	var definitions []map[string]interface{}
+
+	for _, s := range specs {
+		tags = append(tags, s.Tags)
+		paths = append(paths, s.Paths)
+		responses = append(responses, s.Responses)
+		parameters = append(parameters, s.Parameters)
+		definitions = append(definitions, s.Definitions)
+	}
+
+	errs := &multierror.Error{}
+
+	combined := Spec{
+		Swagger:  common.Swagger,
+		Info:     common.Info,
+		BasePath: common.BasePath,
+		Consumes: common.Consumes,
+		Produces: common.Produces,
+		Schemes:  common.Schemes,
+
+		Tags:        combineTags(common.Tags, tags, errs),
+		Paths:       combineSubSpec(common.Paths, paths, "paths", errs),
+		Responses:   combineSubSpec(common.Responses, responses, "responses", errs),
+		Parameters:  combineSubSpec(common.Parameters, parameters, "parameters", errs),
+		Definitions: combineSubSpec(common.Definitions, definitions, "definitions", errs),
+	}
+
+	return combined, errs.ErrorOrNil()
+}
+
+// unmarshalToSwagger converts a list of specs and a common spec
+// from YAML format to Swagger structs.
+// Returned error is a list of errors from incompatible swagger specs.
+func unmarshalToSwagger(yamlCommon string, yamlSpecs []string) (Spec, []Spec, error) {
+	errs := &multierror.Error{}
+
+	editedYAMLSpecs := makeAllYAMLReferencesLocal(yamlSpecs)
+	specs := unmarshalManyFromYAML(editedYAMLSpecs, errs)
+
+	common, err := unmarshalFromYAML(yamlCommon)
+	if err != nil {
+		return Spec{}, nil, err
+	}
+
+	return common, specs, errs.ErrorOrNil()
+}
+
+// unmarshalManyFromYAML maps the passed strings to their respective
+// Swagger specs.
+func unmarshalManyFromYAML(yamlSpecs []string, errs error) []Spec {
+	var specs []Spec
+	for _, swaggerYAML := range yamlSpecs {
```

Comment:
When iterating over a plural, either use (a) the singular form of the plural (`yamlSpec`) or (b) for short loops you can use `val` or a single-char variable name like `s`. In this case, I'd prefer the former, since we're already using a single-char variable name in the loop

Reply:
Thanks for pointing it out, I missed this when I was cleaning up the code